Add support for Hatch Rest noise machine and light

ViViDboarder opened the pull request:
This is a work in progress. Things I have yet to do:

- [x] Get upstream changes merged (or publish fork of package)
    - https://github.com/kjoconnor/pyhatchbabyrest/pull/6
    - https://github.com/kjoconnor/pyhatchbabyrest/pull/5
- [x] Bump dependency version
- [x] Add logo
- [x] Write tests
- [ ] Make platinum🏆

Things I'd appreciate early feedback on:

Entity types. Does it make sense to use a Select and Siren? How should I handle volume? Would a Media Player be better?

<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->
## Breaking change
<!--
  If your PR contains a breaking change for existing users, it is important
  to tell them what breaks, how to make it work again and why we did this.
  This piece of text is published with the release notes, so it helps if you
  write it towards our users, not us.
  Note: Remove this section if this PR is NOT a breaking change.
-->


## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->
This adds a few entities because there doesn't appear to be a great
fitting entity type.

Light: Controls light features
Select: Choose sound setting (because Siren has no UI element for this)
Siren: Turn on sound (also supports selection via service call)
Switch: Turn off the entire device. Can also be accomplished by turning
off both light and sound

What is missing is a UI element to control volume, thought it can be
done via a service call to the Siren.


## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New integration (thank you!)
- [ ] New feature (which adds functionality to an existing integration)
- [ ] Deprecation (breaking change to happen in the future)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue: 
- Link to documentation pull request: 

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [ ] Local tests pass. **Your PR cannot be merged unless tests pass**
- [ ] There is no commented out code in this PR.
- [ ] I have followed the [development checklist][dev-checklist]
- [x] The code has been formatted using Black (`black --fast homeassistant tests`)
- [ ] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [x] Documentation added/updated for [www.home-assistant.io][docs-repository] (https://github.com/home-assistant/home-assistant.io/pull/24120)

If the code communicates with devices, web services, or third-party tools:

- [x] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [x] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.
- [ ] Untested files have been added to `.coveragerc`.

The integration reached or maintains the following [Integration Quality Scale][quality-scale]:
<!--
  The Integration Quality Scale scores an integration on the code quality
  and user experience. Each level of the quality scale consists of a list
  of requirements. We highly recommend getting your integration scored!
-->

- [ ] No score or internal
- [x] 🥈 Silver
- [ ] 🥇 Gold
- [ ] 🏆 Platinum

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [x] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

ViViDboarder commented:
Decided to fork the project. I'll switch back to the main if the maintainer returns and merges.

ViViDboarder commented:
Added a media component as an alternative to Siren and Select because it offers easier volume control. I'll probably remove the other two platforms.

ViViDboarder commented:
Also, I'd like to write tests to cover the actual entities, but I can't seem to get it working as it says it's failing to find the Bluetooth device.

The component that I used as a reference were the Switchbot component, and it doesn't have tests either, possibly due to the same issue. It also uses [`bluetooth.async_ble_device_from_address(...)`](https://github.com/home-assistant/core/blob/dev/homeassistant/components/switchbot/__init__.py#L85) to get the device, which is where mine fails.

Looking at another device that uses Bluetooth discovery, it takes a completely different approach. It uses the [`PassiveBluetoothProcessorCoordinator`](https://github.com/home-assistant/core/blob/dev/homeassistant/components/govee_ble/__init__.py#L30). I'm not sure if this is a preferred approach or only one that works for some types of devices or not, but it does appear to allow testing.

My next option will probably be to mock `bluetooth.async_ble_device_from_address(...)` and see how that goes.

iMicknl reviewed (changes requested):
Please limit the first PR to just one platform (entity). In order to speed up review times, keep the initial PR as small as possible and add new platforms / features in follow-up PR's.

iMicknl reviewed (changes requested):
Adding some quick comments. Please slim down the PR as mentioned in my first comment, so review time will be faster.

ViViDboarder commented on homeassistant/components/hatchrest/__init__.py:
```diff
@@ -0,0 +1,128 @@
+        """Initialize entity coordinator and id."""
+        super().__init__(coordinator)
+        self._device = coordinator.device
+        self._attr_unique_id = coordinator.unique_id
```
Oh, I think that was leftover from before I moved it to the base entity. Soon to push branch will reference `self` now.

They are only really used in one place and, even if more than one component are instantiated, they should share the same memory due to string interning. Is there another reason I should consider this?

ViViDboarder commented on homeassistant/components/hatchrest/__init__.py:
```diff
@@ -0,0 +1,128 @@
+                "model": "Rest",
+            }
+            if self.unique_id
+            else {}
```
It really shouldn't be possible, but the base class has the type `str | None`, so I added this. Alternatively, I could cast or assert, but this feels safer.

rytilahti replied:
Are other bluetooth integrations using this pattern? In any case, `None` is different from `{}`. I would just remove this construct: if the `unique_id` is falsy here, there are some deeper problems and it's better not to hide that.

ViViDboarder replied:
`mypy` doesn't really like that because `self.unique_id` inherits it's type as `str|None`. If I don't add a condition here, I get complaints that `"identifiers": {(DOMAIN, self.unique_id)},` isn't correctly typed because the value `None` is incompatible.

I could resolve this by using an assert or `typing.cast` though. Would that be preferred?

ViViDboarder replied:
If you think it's better to not hide, maybe an assert is best.

rytilahti replied:
I'd simply recommend taking a look at how other similar components do that and following their example on this. I'm not sure why the others are not erroring, as there are plenty of them not doing any sort of special handling. See it yourself by running `ag -A 10 -- 'def device_info' homeassistant/components/`.

ViViDboarder commented on homeassistant/components/hatchrest/__init__.py:
```diff
@@ -0,0 +1,128 @@
+        self._device = coordinator.device
+        self._attr_unique_id = coordinator.unique_id
+
+    @property
+    def name(self):
+        """Name of the Hatch Rest device."""
+        return self._device.name
```
I renamed this property to `device_name` since I use it in the rest of the entity names.

ViViDboarder committed:
Feedback

ViViDboarder committed:
Add support for Hatch Rest noise machine and light

This initially starts with a single switch entity, but there are other
entities forthcoming.

ViViDboarder commented:
Removed all components except the switch. Still looking for some advice on which components to use. I'm thinking probably switch, light, media player.

ViViDboarder commented:
Hello. Do you have any additional feedback? Or any advice on the sound machine part?

ViViDboarder committed:
Run hassfest

ViViDboarder commented:
Thanks. I ran hassfest again. Anything I can do to get this merged? I'm running hass.io and since this uses Discovery, I can't deploy run as a local custom component even. :/

rytilahti commented on homeassistant/components/hatchrest/__init__.py:
```diff
@@ -0,0 +1,122 @@
+import logging
+
+import async_timeout
+from bleak.exc import BleakError
```
```suggestion
from bleak import BleakError
```
It's almost always better to import from the main module if the symbol is exported through it.

ViViDboarder replied:
Is it exported though? My linter (Pyright) tells me that it's not and suggests importing from `bleak.exc` instead. Maybe I have an older version on my branch.

rytilahti replied:
If you look at the other existing components for `BleakError` they seem to import it from the main module. That also worked fine when I tested it in local repl.

rytilahti commented on homeassistant/components/hatchrest/__init__.py:
```diff
@@ -0,0 +1,122 @@
+        """Name of the Hatch Rest device."""
+        return self._device.name
+
+    @callback
+    def _handle_coordinator_update(self) -> None:
+        """Handle updated data from the coordinator."""
+        self.async_write_ha_state()
```
If I'm not mistaken, this is not needed as it's exactly the same what the base class implementation does.

rytilahti commented on homeassistant/components/hatchrest/__init__.py:
```diff
@@ -0,0 +1,122 @@
+        return (
+            {
+                "identifiers": {(DOMAIN, self.unique_id)},
+                "name": self.device_name,
```
This should likely define a connection, too? Something like:
`connections: {(dr.CONNECTION_BLUETOOTH, <mac address>)}`

Also, if the device reports some sort of unique id (e.g., serial number), it would be better as a unique id than the config entry id.

ViViDboarder replied:
I didn't see that about a connection. I can add it.

The `unique_id` here is actually based on the mac address but formatted such that it can be used as a device ID.

rytilahti replied:
It's all bit fuzzy to me, too, so I created an issue to clarify the docs: https://github.com/home-assistant/developers.home-assistant/issues/1571

ViViDboarder committed:
Feedback update

ViViDboarder committed:
Fix flake8 and black

ViViDboarder committed:
Add connection info to device info

ViViDboarder committed:
Switch back to upstream package since my fixes are merged

ViViDboarder commented:
All the tests are failing but it doesn't look related to my change. I can try to pull the dev branch back in or rebase on it again.

ViViDboarder committed:
Add the ability to manually add devices without discovery

rytilahti commented on tests/components/hatchrest/__init__.py:
```diff
@@ -0,0 +1,32 @@
+    BLEDevice("00:00:00:00:00:00", "Hatch Baby Rest"),
+    AdvertisementData(
+        local_name="Hatch Baby Rest",
+        address="00:00:00:00:00:00",
```
I'm not sure about all the tests, but some of them (e.g., https://github.com/home-assistant/core/actions/runs/3696984389/jobs/6261629900) seem to hint towards that this line is the problem:
```
 tests/components/hatchrest/__init__.py:14: in <module>
    AdvertisementData(
E   TypeError: AdvertisementData.__new__() got an unexpected keyword argument 'address'
```
At least the current bleak master branch does not have this argument either: https://github.com/hbldh/bleak/blob/develop/bleak/backends/scanner.py#L22

Maybe fixing that would fix the CI?

ViViDboarder replied:
Looks like the APIs have changed a bit since I wrote it. I'm not noticing tests for anything else using Bluetooth yet, so I'm actually going to remove these tests since they seem flaky and comments on the namedtuples used say the APIs are unstable at the moment.

ViViDboarder replied:
Never mind. I see how the switchbot tests work now and updated mine.

ViViDboarder committed:
Fix tests

rytilahti commented on homeassistant/components/hatchrest/config_flow.py:
```diff
@@ -0,0 +1,63 @@
+        return self.async_create_entry(
+            title=discovery_info.name,
+            data={CONF_ADDRESS: discovery_info.address},
+        )
```
See the automatic warnings above. As discussed at https://developers.home-assistant.io/docs/development_checklist, at least the config flows need to be covered with tests, but it's recommended to have tests for other parts, too.

zacs commented:
So close to the finish line, just a bit of test coverage. Curious @ViViDboarder if this is still on your radar? Appreciate all the work thus far!

ViViDboarder commented:
It is, thanks. My time to work on it is low right now but I hope to get back to it soon. Especially since the official app isn't working so well right now for me.

ViViDboarder commented:
So I know I still need to add the tests, but I encountered some odd behavior when running this as a custom component. 

As I understand, discovery doesn't work for custom components, this is why I added the manual config flow as well. After manually adding a device, I noticed other devices detected by bluetooth polling being automatically added. I think that a Confirm step should be added, so I'll add that to the list of things to do.

The next issue is potentially bigger. When using the integration the clock on my Hatch Rest seemingly starts to drift. The schedules set in the Hatch app shift later and later over time. The problem goes away after syncing clocks, but then starts to manifest again. The odd thing about it is that the clock itself appears to be correct, even though syncing fixes the issue.

This issue can be mitigated by doing all scheduling within Home Assistant, but it isn't great. Particularly when paired with the discovery issue. It could be throwing off schedules for devices nearby as well.

brystmar commented:
Chiming in here as another interested outsider.  I'd be happy to help test with my v1 (bluetooth) Hatch Rest.  Adding this to our toddler's bedtime scene would be amazing!

edenhaus commented on homeassistant/components/hatchrest/translations/en.json:
```diff
@@ -0,0 +1,20 @@
+{
+    "config": {
```
Please remove the translations folder. The translation string to directly in `strings.json`

edenhaus commented on homeassistant/components/hatchrest/switch.py:
```diff
@@ -0,0 +1,51 @@
+    @property
+    def name(self) -> str:
+        """Hatch Rest switch name."""
+        return f"{super().device_name} power"
```
The naming has changed. Please adjust the code to the new [naming requirements](https://developers.home-assistant.io/docs/core/entity#entity-naming)

edenhaus commented on homeassistant/components/hatchrest/__init__.py:
```diff
@@ -0,0 +1,124 @@
+        if not all((self._device.address, self.unique_id)):
+            raise ValueError("Missing bluetooth address for hatch rest device")
+
+        assert self._device.address
+        assert self.unique_id
```
This seems like a bug as you already checking if both are set above.

This is a great example of asserting should not be used. Typing is here to help, asserting seems easy, but actually (in most cases) means either something is wrong, or typing was wrong to being with.

In this case, there is something wrong, which is now masked by asserting. Defeating the purpose of typing it to being with.

edenhaus commented on homeassistant/components/hatchrest/__init__.py:
```diff
@@ -0,0 +1,124 @@
+    def device_info(self) -> DeviceInfo:
+        """Provide info about the Hatch Rest device for all entities."""
+        # These should never be null or empty
+        if not all((self._device.address, self.unique_id)):
```
Where is `self.unique_id` set?

edenhaus commented on homeassistant/components/hatchrest/switch.py:
```diff
@@ -0,0 +1,51 @@
+        """Turn on the Hatch Rest device."""
+        if not self.is_on:
+            await self._device.power_on()
+            self._device.power = True
```
```suggestion
```

I would set the power state directly in `power_on` function and make the property `power` read only.
So it is not possible (by mistake) do only set `power=True` without even calling `power_on`

edenhaus commented on homeassistant/components/hatchrest/switch.py:
```diff
@@ -0,0 +1,51 @@
+        """Turn off the Hatch Rest device."""
+        if self.is_on:
+            await self._device.power_off()
+            self._device.power = False
```
```suggestion
```

Same here

edenhaus reviewed (changes requested):
Thanks for your contribution :)  I have done a quick first review


As @rytilahti already noted, we require 100% test coverage for the config flow. More tests are always welcome :)
Please rebase the PR to get the latest changes.
